pkg/store/bucket.go

```diff
@@ -2325,112 +2318,232 @@ func ExpandPostingsWithContext(ctx context.Context, p index.Postings) (res []sto
 	return res, p.Err()
 }
 
-// postingGroup keeps posting keys for single matcher. Logical result of the group is:
+// postingGroup keeps posting keys for one or more matchers with the same label name. Logical result of the group is:
 // If addAll is set: special All postings minus postings for removeKeys labels. No need to merge postings for addKeys in this case.
 // If addAll is not set: Merge of postings for "addKeys" labels minus postings for removeKeys labels
 // This computation happens in ExpandedPostings.
 type postingGroup struct {
 	addAll     bool
-	addKeys    []labels.Label
-	removeKeys []labels.Label
+	name       string
+	addKeys    []string
+	removeKeys []string
 }
 
-func newPostingGroup(addAll bool, addKeys, removeKeys []labels.Label) *postingGroup {
+func newPostingGroup(addAll bool, name string, addKeys, removeKeys []string) *postingGroup {
 	return &postingGroup{
 		addAll:     addAll,
+		name:       name,
 		addKeys:    addKeys,
 		removeKeys: removeKeys,
 	}
 }
 
-func checkNilPosting(l labels.Label, p index.Postings) index.Postings {
+func (pg postingGroup) merge(other *postingGroup) *postingGroup {
+	if other == nil {
+		return &pg
+	}
+	// This shouldn't happen, but add this as a safeguard.
+	if pg.name != other.name {
+		return nil
+	}
+	var i, j int
+	// Both add all, merge remove keys.
+	if pg.addAll && other.addAll {
+		output := make([]string, 0, len(pg.removeKeys)+len(other.removeKeys))
+		for i < len(pg.removeKeys) && j < len(other.removeKeys) {
+			if pg.removeKeys[i] < other.removeKeys[j] {
+				output = append(output, pg.removeKeys[i])
+				i++
+			} else if pg.removeKeys[i] > other.removeKeys[j] {
+				output = append(output, other.removeKeys[i])
+				j++
+			} else {
+				output = append(output, pg.removeKeys[i])
+				i++
+				j++
+			}
+		}
+		for i < len(pg.removeKeys) {
+			output = append(output, pg.removeKeys[i])
```

Comment:
To make it easier I think we can call `output = append(output, pg.removeKeys[i:]...)`